internal/provider/auth_token_file.go

```diff
@@ -72,7 +71,17 @@ func (l *AuthLoginTokenFile) Init(d *schema.ResourceData,
 ) (AuthLogin, error) {
 	l.mount = consts.MountTypeNone
 
+	defaults := authDefaults{
+		{
+			field:      consts.FieldFilename,
+			envVar:     consts.EnvVarTokenFilename,
+			defaultVal: "",
+		},
+	}
 	if err := l.AuthLoginCommon.Init(d, authField,
+		func(data *schema.ResourceData) error {
+			return l.setDefaultFields(d, defaults)
```

Comment:
Nice! I think this is a great clean way to handle defaults ourselves and gain more control over them. I do not foresee any issues of handling defaults this way since we can ensure the behavior works as expected by using/enhancing our current test cases.

With the updates made to `checkRequiredFields`, we are now also ensuring that the field is only 'missing' if it is not provided by the user and not set by the environment variable, which makes sense and mirrors the behavior of today. 

I am good with these changes on how we are handling defaults and required fields in the Auth resources. 👍🏼

Reply:
Wanted to note a small follow-up after looking at the failing tests — This does slightly change the expected return value of `l.params` when we fail any of the validations below. I see we moved [these lines](https://github.com/hashicorp/terraform-provider-vault/pull/2090/files#diff-5a37f9a7fa8158cc2773750fe9bbc5c7766c3a78bf1a6cb6778d3a76df511ccaR146-R147) up in order to ensure that `l.params` is non-nil so that we can make use of it in `setDefaultFields` and `checkRequiredFields`. This would imply that before, when we expected `l.params` to be nil on validation, we would now expect it to be populated with the entire config with zero values,  as demonstrated by this [line in the CI](https://github.com/hashicorp/terraform-provider-vault/actions/runs/6909740801/job/18801634607?pr=2090#step:7:148). I don't think this is necessarily a problem, just something to be mindful of

I wonder if instead of pre-populating `l.params` before the validation is complete, if setting `l.params = map[string]interface{}{}` [on this line in](https://github.com/hashicorp/terraform-provider-vault/blob/VAULT-21288/auth-login-fwprovider/internal/provider/auth.go#L275) `init()` would achieve the same purpose 🤔 This way, we won't populate all the fields in `l.params`, only with the defaults, and will still be able to make use of it in our validators.